File: UIWidgetTest/app/src/main/java/com/banshion/uiwidgettest/MainActivity.java
package com.banshion.uiwidgettest;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.Toast;

import com.banshion.uiwidgettest.fragment.AnotherRightFragment;
import com.banshion.uiwidgettest.fragment.RightFragment;
import com.banshion.uiwidgettest.recycler.FruitAdapter;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements View.OnClickListener{
    private List<Fruit> fruitList = new ArrayList<>();
    private String[] data = {"Apple", "Banana", "Orange", "Watermelon", "Pear", "Grape", "Pineapple", "Strawberry", "Cherry", "Mango",
            "Apple", "Banana", "Orange", "Watermelon", "Pear", "Grape", "Pineapple", "Strawberry", "Cherry", "Mango"};
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ActionBar supportActionBar = getSupportActionBar();
        if (supportActionBar != null) {
            supportActionBar.hide();
        }
        Button button = findViewById(R.id.button);
        button.setOnClickListener(this);
        replaceFragment(new RightFragment());
//        initFruit();
//        RecyclerView recyclerView = findViewById(R.id.recycle_view);
//        StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(3, StaggeredGridLayoutManager.VERTICAL);
//        recyclerView.setLayoutManager(staggeredGridLayoutManager);
//        FruitAdapter fruitAdapter = new com.banshion.uiwidgettest.recycler.FruitAdapter(fruitList);
//        recyclerView.setAdapter(fruitAdapter);
//        ListView listView = findViewById(R.id.list_view);
//        ArrayAdapter<String> stringArrayAdapter = new ArrayAdapter<>(MainActivity.this, android.R.layout.simple_list_item_1, data);
//        listView.setAdapter(stringArrayAdapter);
//        FruitAdapter fruitAdapter = new FruitAdapter(MainActivity.this, R.layout.fruit_item, fruitList);
//        listView.setAdapter(fruitAdapter);
//        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
//            @Override
//            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
//                Fruit fruit = fruitList.get(position);
//                Toast.makeText(MainActivity.this, fruit.getName(), Toast.LENGTH_SHORT).show();
//            }
//        });

    }

    private void initFruit() {
        for (int i = 0; i < data.length; i++) {
            Fruit fruit = new Fruit(data[i], R.drawable.apple_pic);
            fruitList.add(fruit);
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.button:
                replaceFragment(new AnotherRightFragment());
                break;
        }
    }

    public void replaceFragment(Fragment fragment) {
        FragmentManager supportFragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = supportFragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.right_fragment, fragment);
        fragmentTransaction.addToBackStack(null);//addToBackStack()参数用于收一个名字用于描述返回栈的状态，一般传入null
        fragmentTransaction.commit();
    }
}

File: BroadcastTestKotlin/app/src/main/java/com/banshion/broadcasttestkotlin/MainActivity.kt
package com.banshion.broadcasttestkotlin

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    lateinit var timerChangeReceiver: TimeChangeReceiver

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val intentFilter = IntentFilter()
        //监听系统时间变化广播
        intentFilter.addAction("android.intent.action.TIME_TICK")
        timerChangeReceiver = TimeChangeReceiver()
        registerReceiver(timerChangeReceiver, intentFilter)

        sendBroadcast.setOnClickListener{
            val intent=Intent("com.example.broadcasttest.MY_BROADCAST")
            /**
             * 在Android 8.0系统之后，静态注册的BroadcastReceiver是无法接收隐式广播的，
             * 而默认情况下我们发出的自定义广播恰恰都是隐式广播。
             * 因此这里一定要调用setPackage()方法，指定这条广播是发送给哪个应用程序的，从而让它变成一条显式广播，
             * 否则静态注册的BroadcastReceiver将无法接收到这条广播。
             */
            intent.setPackage(packageName)
//            sendBroadcast(intent)发送标准广播
            //发送有序广播第一个参数仍然是Intent；第二个参数是一个与权限相关的字符串，这里传入null就行了
            sendOrderedBroadcast(intent, null)
        }

    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(timerChangeReceiver)
    }

    inner class TimeChangeReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            Toast.makeText(context, "Time has changed", Toast.LENGTH_SHORT).show()
        }

    }
}